### ISSUE TITLE
Binary Classification Results

### ISSUE THREAD

**Issue by lefig** [dotnet/machinelearning#737]
Hi all,

I have a quick question and would appreciate your thoughts regarding what could be a dubious score. I have been training/testing some bin classification models and the results look too promising. For instance,

Not training a calibrator because it is not needed.
*************************************************
*       Metrics for Logistic Regression Binary
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%

This is not a software issue but rather a query. I would have expected accuracy to have been reported to decimal places (less than 1).

Thanks

Fig

**Comment by rogancarr** [comment]
Hi Fig,

It looks like you are getting perfect accuracy, hence a `1`. Are you testing on the train set or perhaps leaking a label in your training data?

**Comment by lefig** [comment]
Good point. Yes, I have separated the train/test data and was studying the results. The predicted results seem true but I was hoping to select the "best" trainer  based on a highest score. ...

Training calibrator.
*************************************************
*       Metrics for Perception
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************
Not adding a normalizer.
Making per-feature arrays
Changing data from row-wise to column-wise
Processed 518 instances
Binning and forming Feature objects
Reserved memory for tree learner: 457080 bytes
Starting to train ...
Training calibrator.
*************************************************
*       Metrics for Fast Forest Binary
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************
Automatically adding a MinMax normalization transform, use 'norm=Warn' or 'norm=No' to turn this behavior off.

Training calibrator.
*************************************************
*       Metrics for Linear SVM
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************
Automatically adding a MinMax normalization transform, use 'norm=Warn' or 'norm=No' to turn this behavior off.
LBFGS multi-threading will attempt to load dataset into memory. In case of out-of-memory issues, add 'numThreads=1' to the trainer arguments and 'cache=-' to the command line arguments to turn off multi-threading.
Beginning optimization
num vars: 15
improvement criterion: Mean Improvement
L1 regularization selected 2 of 15 weights.
Not training a calibrator because it is not needed.
*************************************************
*       Metrics for Logistic Regression Binary
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************
Automatically adding a MinMax normalization transform, use 'norm=Warn' or 'norm=No' to turn this behavior off.
Using 2 threads to train.
Automatically choosing a check frequency of 2.
Auto-tuning parameters: maxIterations = 2894.
Auto-tuning parameters: L2 = 2.670127E-05.
Auto-tuning parameters: L1Threshold (L1/L2) = 0.
Using best model from iteration 62.
Not training a calibrator because it is not needed.
*************************************************
*       Metrics for Stochastic Dual Coordinate Ascent Binary
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************
Not adding a normalizer.
Making per-feature arrays
Changing data from row-wise to column-wise
Processed 518 instances
Binning and forming Feature objects
Reserved memory for tree learner: 105480 bytes
Starting to train ...
Not training a calibrator because it is not needed.
*************************************************
*       Metrics for Fast Tree Binary
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************

**Comment by lefig** [comment]
Good point. Yes, I have separated the train/test data and was studying the results. The predicted results seem true but I was hoping to select the "best" trainer  based on a highest score. ...

Training calibrator.
*************************************************
*       Metrics for Perception
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************
Not adding a normalizer.
Making per-feature arrays
Changing data from row-wise to column-wise
Processed 518 instances
Binning and forming Feature objects
Reserved memory for tree learner: 457080 bytes
Starting to train ...
Training calibrator.
*************************************************
*       Metrics for Fast Forest Binary
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************
Automatically adding a MinMax normalization transform, use 'norm=Warn' or 'norm=No' to turn this behavior off.

Training calibrator.
*************************************************
*       Metrics for Linear SVM
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************
Automatically adding a MinMax normalization transform, use 'norm=Warn' or 'norm=No' to turn this behavior off.
LBFGS multi-threading will attempt to load dataset into memory. In case of out-of-memory issues, add 'numThreads=1' to the trainer arguments and 'cache=-' to the command line arguments to turn off multi-threading.
Beginning optimization
num vars: 15
improvement criterion: Mean Improvement
L1 regularization selected 2 of 15 weights.
Not training a calibrator because it is not needed.
*************************************************
*       Metrics for Logistic Regression Binary
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************
Automatically adding a MinMax normalization transform, use 'norm=Warn' or 'norm=No' to turn this behavior off.
Using 2 threads to train.
Automatically choosing a check frequency of 2.
Auto-tuning parameters: maxIterations = 2894.
Auto-tuning parameters: L2 = 2.670127E-05.
Auto-tuning parameters: L1Threshold (L1/L2) = 0.
Using best model from iteration 62.
Not training a calibrator because it is not needed.
*************************************************
*       Metrics for Stochastic Dual Coordinate Ascent Binary
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************
Not adding a normalizer.
Making per-feature arrays
Changing data from row-wise to column-wise
Processed 518 instances
Binning and forming Feature objects
Reserved memory for tree learner: 105480 bytes
Starting to train ...
Not training a calibrator because it is not needed.
*************************************************
*       Metrics for Fast Tree Binary
*------------------------------------------------
*       Accuracy: 1
*       Entropy: 0.999316959634874
*       Auc: 100.00%
*       F1Score: 100.00%
*************************************************

**Comment by rogancarr** [comment]
Hi Fig,

Usually, you'll see a spread of values for metrics like Accuracy, AUC, F1, etc., as you try out different learners. As to why accuracy is a `1`, it's because it's getting a perfect score, and the print statement is simply printing `1` rather than a fixed number of decimal places.

Now, as to why this is happening, you are probably leaking a label in your training set. That is, there is some column that corresponds exactly to the label, and each model is learning exactly that label. To debug this, I'd take a look at the `summary` of the models to see which features are being given the most weight.

**Comment by lefig** [comment]
Hi Rogan,

Thank you for looking into this. 

Ah, that explains things somewhat. I will certainly look into the summary of the models. 

When you mention about label leaking, now that does sound interesting. Apologies, but what does that actually mean...? That a feature has been incorrectly modelled perhaps.

Thanks

**Comment by rogancarr** [comment]
Hi Fig,

Label leaking means that you have a feature which is actually the label in disguise. Imagine that you are trying to predict whether an item was ever purchased in an online store, and one of the features used is `price`. However, let's say that your logging system only logs the `price` if the item was sold. Then the `price` feature will have a numerical value for all those sold, and a `null` for those which were unsold. This feature now actually encodes the label directly, so a learning algorithm can just look at the value of `price` to exactly determine the label, ignoring all other features. The model produced by learning on this dataset, however, is worthless because all the unseen data in production will have a price regardless of its purchase status. This type of scenario is called label leaking.

**Comment by lefig** [comment]
How right you were too! Yes, it looks like I was incorrectly including the target class as a feature. Now I am looking at the results and am surprised to see how poor the results are. But I was wondering whether you may know why I am unable to capture all metrics such as
*************************************************
*       Metrics for Generalized Additive Model
*------------------------------------------------
*       R2 Score: 0.84
*       Absolute loss:
*       Squared loss:
*       RMS loss:
*************************************************

**Comment by codemzs** [comment]
please refer to samples here https://github.com/dotnet/machinelearning/tree/master/docs/samples/Microsoft.ML.Samples/Dynamic

also closing since this issues seems to have been addressed.
